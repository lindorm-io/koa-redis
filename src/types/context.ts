import { KoaContext } from "@lindorm-io/koa";
import { RedisCache, RedisConnection } from "@lindorm-io/redis";

export interface RedisContext extends KoaContext {
  cache: Record<string, RedisCache>;
  client: { redis: RedisConnection };
}
