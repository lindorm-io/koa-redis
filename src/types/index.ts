import { IKoaAppContext } from "@lindorm-io/koa";
import { IRedisConnectionOptions, RedisConnection, TRedisClient } from "@lindorm-io/redis";

export interface IRedisMiddlewareOptions extends IRedisConnectionOptions {
  clientRef?: (client: TRedisClient) => void;
}

export interface IRedisMiddlewareContext extends IKoaAppContext {
  redis: RedisConnection;
}
