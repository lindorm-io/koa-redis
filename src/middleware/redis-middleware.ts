import { IRedisMiddlewareContext, IRedisMiddlewareOptions } from "../types";
import { RedisConnection, RedisConnectionType } from "@lindorm-io/redis";
import { TPromise } from "@lindorm-io/core";

export const redisMiddleware = (options: IRedisMiddlewareOptions) => async (
  ctx: IRedisMiddlewareContext,
  next: TPromise<void>,
): Promise<void> => {
  const start = Date.now();

  ctx.redis = new RedisConnection(options);

  await ctx.redis.connect();

  if (options.type === RedisConnectionType.MEMORY && options.clientRef) {
    options.clientRef(ctx.redis.getClient());
    ctx.logger.debug("redis client ref called");
  }

  ctx.logger.debug("redis connection established");

  ctx.metrics = {
    ...(ctx.metrics || {}),
    redisConnection: Date.now() - start,
  };

  try {
    await next();
  } finally {
    if (options.type !== RedisConnectionType.MEMORY) {
      await ctx.redis.disconnect();
    }
  }
};
