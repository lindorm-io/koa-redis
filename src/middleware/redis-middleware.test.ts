import MockDate from "mockdate";
import { redisMiddleware } from "./redis-middleware";
import { RedisConnection, RedisConnectionType, RedisInMemoryClient, TRedisClient } from "@lindorm-io/redis";
import { Logger, LogLevel } from "@lindorm-io/winston";
import { IRedisMiddlewareOptions } from "../types";

jest.mock("uuid", () => ({
  v4: () => "e397bc49-849e-4df6-a536-7b9fa3574ace",
}));

MockDate.set("2020-01-01 08:00:00.000");

const logger = new Logger({ packageName: "n", packageVersion: "v" });
logger.addConsole(LogLevel.ERROR);

describe("redisMiddleware", () => {
  let client: TRedisClient;
  let options: IRedisMiddlewareOptions;
  let ctx: any;
  let next: any;

  beforeEach(() => {
    options = {
      type: RedisConnectionType.MEMORY,
      port: 100,
      clientRef: (c) => {
        client = c;
      },
    };
    ctx = { logger };
    next = () => Promise.resolve();
  });

  test("should set a functional redis on context", async () => {
    await expect(redisMiddleware(options)(ctx, next)).resolves.toBe(undefined);

    expect(ctx.redis).toStrictEqual(expect.any(RedisConnection));
    expect(client).toStrictEqual(expect.any(RedisInMemoryClient));

    await expect(client.set("key", { blob: "yes" })).resolves.toBe("OK");
    await expect(client.get("key")).resolves.toMatchSnapshot();
  });
});
