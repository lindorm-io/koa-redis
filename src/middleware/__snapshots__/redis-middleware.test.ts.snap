// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redisMiddleware should set a functional redis on context 1`] = `
Object {
  "blob": "yes",
}
`;
